beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+    auto masked_weights = ranges::views::zip(weights, mask) |   //
+                          ranges::views::filter(mask_filter) |  //
+                          beluga::views::elements<0>;           //
+
+    const auto& [total_weight, particle_count] = info;
+
+    // if there's only one sample in the cluster we can't estimate the covariance
+    if (particle_count < 2) {
+      continue;
+    }
+
+    const auto [mean, covariance] = beluga::estimate(masked_states, masked_weights);
+    cluster_estimates.emplace_back(total_weight, std::move(mean), std::move(covariance));
+  }
+
+  return cluster_estimates;
+}
+
+}  // namespace cse_detail
+
+/// Primary template for a cluster-based estimation algorithm.
+/**
+ * Particles are groups into clusters around local maximums and the mean and covariance of the cluster with the highest
+ * weight is returned.
+ *
+ * This class implements the EstimationInterface interface
+ * and satisfies \ref StateEstimatorPage.
+ * *
+ * \tparam Mixin The mixin that implements the particle filter interface.
+ * \tparam StateType The type of the state to be estimated.
+ * \tparam CovarianceType The type of the covariance matrix of the state to be estimated.
+ */
+template <class Mixin, class StateType, class CovarianceType>
+class ClusterBasedStateEstimator : public Mixin {
+ public:
+  /// Constructs a ClusterBasedStateEstimator instance.
+  /**
+   * \tparam ...Args Arguments types for the remaining mixin constructors.
+   * \param ...args Arguments that are not used by this part of the mixin, but by others.
+   */
+  template <class... Args>
+  explicit ClusterBasedStateEstimator(Args&&... args) : Mixin(std::forward<Args>(args)...) {}
+
+  /// \copydoc EstimationInterface2d::estimate()
+  [[nodiscard]] std::pair<StateType, CovarianceType> estimate() const final;
+
+ private:
+  static constexpr double kSpatialHashResolution = 0.20;   ///< clustering algorithm spatial resolution
+  static constexpr double kAngularHashResolution = 0.524;  ///< clustering algorithm angular resolution
+
+  /// @brief spatial hash function used to group particles in cells
+  const beluga::spatial_hash<Sophus::SE2d> spatial_hash_function_{
+      kSpatialHashResolution,  // x
+      kSpatialHashResolution,  // y
+      kAngularHashResolution   // theta
+  };
+
+  /// @brief Adjacent grid cells used by the clustering algorithm.
+  const std::vector<Sophus::SE2d> adjacent_grid_cells_ = {
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{+kSpatialHashResolution, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{-kSpatialHashResolution, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{0.0, +kSpatialHashResolution}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{0.0, -kSpatialHashResolution}},
+      Sophus::SE2d{Sophus::SO2d{+kAngularHashResolution}, Sophus::Vector2d{0.0, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{-kAngularHashResolution}, Sophus::Vector2d{0.0, 0.0}},
+  };
+};
```

Comment:
@glpuga meta: considering that both spatial hashing / indexing and space discretization must have matching resolutions, it would be nice for local neighbor cell poses to be derived from the given spatial hashing / indexing mechanism (or otherwise deduplicated somehow).

Reply:
I'm not sure I understand this comment.